src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+    let five_seconds : unit Lwt.t = Lwt_unix.sleep 5. in
+
+    (* Both waits have already been started at this point! *)
+
+    Lwt.bind three_seconds (fun () ->
+      (* This is 3 seconds later. *)
+      Lwt.bind (Lwt_io.printl "3 seconds passed") (fun () ->
+        Lwt.bind five_seconds (fun () ->
+          (* Only 2 seconds were left in the 5-second wait, so
+              this callback runs 2 seconds after the first callback. *)
+          Lwt_io.printl "Only 2 more seconds passed")))
+  end
+
+(* ocamlfind opt -linkpkg -package lwt.unix code.ml && ./a.out *)
+]}
+
+    And that's it! Concurrency in Lwt is simply a matter of whether you start an
+    operation in the callback of another one or not.
+
+    {3 Execution model}
+
+    It's important to understand that promises are a pure-OCaml data type. They
+    don't do any fancy scheduling or I/O. They are just lists of callbacks if
+    pending, and containers for one value if resolved.
+
+    The interesting function is {!Lwt_main.run}. It's a wrapper around
+    {{: http://man7.org/linux/man-pages/man2/select.2.html} [select(2)]},
+    {{: http://man7.org/linux/man-pages/man7/epoll.7.html} [epoll(7)]},
+    {{: https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2}
+    [kqueue(2)]}, or whatever asynchronous API your system provides. Let's take
+    [select(2)] for the example.
+
+    On browsers, the [select(2)] call is inside the JavaScript engine, so you
+    don't call {!Lwt_main.run}. But the logic is still the same!
+
+    Let's use the program from the tutorial that reads two lines as an example.
+    Here it is, again, in its desugared form:
+
+{[
+let () =
+  let p : unit Lwt.t =
+    let line_1_promise : string Lwt.t = Lwt_io.(read_line stdin) in
+    Lwt.bind line_1_promise (fun (line_1 : string) ->
+
+      let line_2_promise : string Lwt.t = Lwt_io.(read_line stdin) in
+      Lwt.bind line_2_promise (fun (line_2 : string) ->
+
+        Lwt_io.printf "%s and %s\n" line_1 line_2))
+  in
+
+  Lwt_main.run p
+
+(* ocamlfind opt -linkpkg -package lwt.unix code.ml && ./a.out *)
+]}
+
+    {!Lwt_main.run} is your program's main I/O loop. You pass it a single
+    top-level promise, and it:
+
+    + Uses [select(2)] to put your process to sleep until the next I/O
+      completes.
+    + That next I/O happens to be the one that reads [line_1]. {!Lwt_main.run}
+      knows that I/O is supposed to resolve [line_1_promise], so it puts
+      [line_1] into the promise and resolves it.
+    + Resolving is an ordinary OCaml operation. It causes all the callbacks of
+      [line_1_promise] to run, one after another. Each callback is also ordinary
+      OCaml code. In our case, there is only one callback, but in general, there
+      might be several, and they might also resolve additional promises. So,
+      promise resolution triggers a “cascade” of callbacks. Eventually, however,
+      we should run out of callbacks, and control will return to
+      {!Lwt_main.run}.
```

Comment:
> Also, another thing it should say is that this example is a simple promise with a string of callbacks. However, in more complex programs with non-trivial arrangements of callbacks/promises, instead of putting the process to sleep when it reaches an I/O operation, the scheduler will first execute callbacks that are ready. Only when it runs out of those does it select.

Actually, the scheduler (there is none, see [comment](https://github.com/ocsigen/lwt/pull/469#issuecomment-333222702), so I mean `Lwt_main.run`) doesn't execute any callbacks that are ready. They are all executed implicitly by promise resolution, by callbacks triggering each other, not by a scheduler. So this paragraph is that additional paragraph you suggest to write, but I guess it should be clarified, and maybe the `Lwt_main.run` paragraph should be split off and/or reordered.

Reply:
With this conversation, I'm realising how bad I am at explaining Lwt… Which makes me realise: Good work on the whole thing!